Discharge temperature

sophieoneill opened the pull request:
This PR adds the discharge temperature functionality to pySimdeum. Added functions for calculating and aggregating discharge temperatures (`assign_discharge_temperature` and `hh_discharge_temperature`). 

Enhancements to discharge data calculations:

* [`pysimdeum/core/end_use.py`](diffhunk://#diff-711b848990e395b81dccb7369ea306ddb9f14b07284731ebc58633ee66732446R179): Added `discharge_temperature` to the discharge event data, updated the calculation logic to handle cycle times, and modified the structure of `start` and `end` times to be lists of cycle times. 

Utility functions for post-processing:

* [`pysimdeum/core/utils.py`](diffhunk://#diff-79d2f6dca9e2370bf217fbef6df6f405f25597dfca160446eb800c7533fb7ac4R431-R471): Introduced a new helper function `discharge_postprocessing` for post-processing discharge data, and refactored nutrient and temperature assignment functions to use this helper. 

Configuration updates:

* Added `discharge_temperature` to all end uses. Add in the subtypes where they exist.

Note that temperature discharged will change over time so this does not give a complete accurate value

val-ismaili commented on pysimdeum/data/UK/end_uses/WashingMachine.toml:
I thought washing machines have a cold rinse cycle at the end and the warm water is one of the earlier cycles

val-ismaili commented on pysimdeum/data/NL/end_uses/WashingMachine.toml:
```diff
@@ -2,6 +2,7 @@
 classname = 'WashingMachine'
 offset = 0
 temperature = 10
+discharge_temperature = [35, 35, 35, 45]
```
see comment on UK config file

val-ismaili commented on pysimdeum/data/NL/end_uses/Dishwasher.toml:
```diff
@@ -2,6 +2,7 @@
 classname = 'Dishwasher'
 offset = 0
 temperature = 10
+discharge_temperature = [35, 0, 0, 0]
```
see comment on uk config file

val-ismaili commented on pysimdeum/core/utils.py:
```diff
@@ -559,6 +617,86 @@ def hh_discharge_nutrients(ds, time_agg='h'):
+        pd.DataFrame: The updated DataFrame containing the discharge data and the temperatures.
+
+    """
+    def calculate_temperature(event_data, total_flow):
```
I don't like the nested function, seems unnecessary. It becomes hard to write unit tests for (when we come to that).

val-ismaili commented on pysimdeum/core/utils.py:
```diff
@@ -436,65 +487,41 @@ def assign_discharge_nutrients(ds):
 
-    df, ref_start, ref_end = xarray_to_metadata_df(ds, 'discharge', 'discharge_events')
-
+    def calculate_nutrients(event_data, total_flow):
```
same here: I don't like the nested function, seems unnecessary. It becomes hard to write unit tests for (when we come to that).

val-ismaili commented on pysimdeum/core/utils.py:
```diff
@@ -418,6 +428,47 @@ def truncated_normal_dis_sampling(mean_value):
+
+    Args:
+        ds (xarray.Dataset): The dataset containing discharge data and discharge events metadata.
+        process_type (str): The type of post-processing to perform.
```
`process_type` isn't used anywhere is it?

val-ismaili commented on pysimdeum/core/utils.py:
```diff
@@ -418,6 +428,47 @@ def truncated_normal_dis_sampling(mean_value):
+        total_flow = event_data['flow'].sum()
+
+        # Process speific calculation
+        values = process_calculation(event_data, total_flow)
```
if you want to do it by this method, i think its still better to take the nested function out of their parent functions so that they can be tested. its fine to pass them as an argument into another function though

sophieoneill commented:
@val-ismaili I've removed the nested functions, as I don't want them to interfere with unit tests. The shared function now does the calculations previously done in the nested functions depending on the `process_type` provided. The nutrient and temperature functions just call this post processing function.
 
I have also updated the `WashingMachine` and `Dishwasher` `discharge_temperature`, after conversations with Olivia, so that for every cycle it samples a value between 30 - 60 and 40 - 60 respectively.

Spot check nutrients, looks ok still.

Temperature function can be run similar to nutrients:
`utils.hh_discharge_temperature(discharge, time_agg='30min')`
This will output a dataframe. Sample below of the temperature aggregated to 30 min as well as the associated aggregated flow:

![image](https://github.com/user-attachments/assets/6ba7d048-fc2d-4d71-bb69-6a58876d39eb)

![image](https://github.com/user-attachments/assets/d2a139f4-dded-4efe-bde1-1fd872804c52)

val-ismaili reviewed (changes requested):
The function `assign_discharge_temperature()` has no purpose. you're just using it as a way to take args into another function and return that functions return values. Just do this directly...In line 673 you can replace that with line 641 and get rid of `assign_discharge_temperature()`

sophieoneill commented:
@val-ismaili That change has been made, and docs have been updated: [Docs](https://pysimdeum--83.org.readthedocs.build/en/83/discharge/water_quality/)
